slice/internal/controller/workload_controller.go

```diff
@@ -179,10 +178,138 @@ func shouldFinalize(wl *kueue.Workload) (bool, string) {
 		return true, "it is no longer active"
 	}
 
+	if !controllerutil.HasControllerReference(wl) {
+		return true, "it doesn't have owner"
+	}
+
+	if !hasSupportedOwner(wl) {
+		return true, "it has an unsupported owner"
+	}
+
 	return false, ""
 }
 
+func hasSupportedOwner(wl *kueue.Workload) bool {
+	// For now, we only support JobSets.
+	return isJobSetOwner(wl)
+}
+
+func isJobSetOwner(wl *kueue.Workload) bool {
+	if owner := metav1.GetControllerOf(wl); owner != nil {
+		return owner.APIVersion == jobset.SchemeGroupVersion.String() && owner.Kind == "JobSet"
+	}
+	return false
+}
+
+func (r *WorkloadReconciler) cleanupSlice(ctx context.Context, wl *kueue.Workload) (bool, error) {
+	slice := v1alpha1.Slice{}
+	sliceKey := core.SliceKeyFromWorkload(wl)
+
+	log := ctrl.LoggerFrom(ctx).WithValues("slice", klog.KRef(sliceKey.Namespace, sliceKey.Name))
+	ctrl.LoggerInto(ctx, log)
+
+	err := r.client.Get(ctx, sliceKey, &slice)
+	if apierrors.IsNotFound(err) {
+		// slice not found
+		return true, nil
+	} else if err != nil {
+		// error fetching slice
+		log.Error(err, "Failed to fetch the Slice")
+		return false, err
+	}
+
+	if !slice.DeletionTimestamp.IsZero() {
+		return core.Deformed(&slice), nil
+	}
+
+	terminated, err := r.ownerPodsFinished(ctx, wl)
+	if err != nil || !terminated {
+		return false, err
+	}
+
+	log.V(3).Info("All Pods in the Workload have finished. Deleting the Slice")
+
+	err = r.client.Delete(ctx, &slice)
+	if apierrors.IsNotFound(err) {
+		return true, nil
+	} else if err != nil {
+		log.Error(err, "Failed to delete the Slice")
+	}
+
+	return false, err
```

Comment:
is this correct? in the normal delete case, we don't expect an error, right?
```suggestion
	if err != nil && !apierrors.IsNotFound {
		log.Error(err, "Failed to delete the Slice")
		return false, err
	}
	return true, nil
```

Reply:
This is because we still need to wait for the status to become 'Deformed' and for the Pods to terminate.